scheduler/src/cook/mesos/task.clj

```diff
@@ -0,0 +1,167 @@
+(ns cook.mesos.task
+  (:require [clojure.algo.generic.functor :refer (fmap)]
+            [clojure.tools.logging :as log])
+  (import com.netflix.fenzo.TaskAssignmentResult))
+
+
+(defn TaskAssignmentResult->task-info
+  "Organizes the info Fenzo has already told us about the task we need to run"
+  [^TaskAssignmentResult fenzo-result]
+  (merge (:task-info (.getRequest fenzo-result))
+         {:ports-assigned (.getAssignedPorts fenzo-result)}))
+
+
+(defmulti combine-like-resources
+  (fn [list] (-> list first :type)))
+
+(defmethod combine-like-resources :value-scalar [resources]
+  (->> resources (map :scalar) (apply +)))
+
+(defmethod combine-like-resources :value-ranges [resources]
+  (->> resources (map :ranges) (apply concat) vec))
+
+(defn resources-by-role
+  "Given a set of offers, combine all the available resources into a comprehensible
+  and easily walkable data strucuture, grouped first by resource name e.g. mem, then
+  by role."
+  [offers]
+  (->> offers
+       (map :resources)
+       flatten
+       (group-by :name)
+       (fmap (fn [resources]
+               (->> resources
+                    (group-by :role)
+                    (fmap combine-like-resources))))))
+
+(defn take-resources
+  "Given a set of available resources (in the format returned by resources-by-role),
+  take the specified amount of the specified resource from the pool.
+  Return {:remaining-resources (the pool after resources were taken)
+          :mesos-messages (the Mesos messages that are necessary to reserve the resources)
+          :amount-still-needed (the amount still needed to satisfy amount... if it's not 0, something went wrong)}"
+  [available-resources resource-name amount]
+  (let [avail (available-resources resource-name)
+        ;; to be polite to other frameworks, take from role-specific pool first.
+        sorted-roles (sort-by #(= "*" %) (keys avail))
+        init-state {:remaining-resources avail
+                    :mesos-messages []
+                    :amount-still-needed amount}]
+    (reduce (fn [{:keys [remaining-resources mesos-messages amount-still-needed]
+                  :as state} role-name]
+              (let [amount-avail (or (remaining-resources role-name) 0)
+                    amount-to-take (min amount-still-needed amount-avail)]
+                (if (pos? amount-to-take)
```

Comment:
@wyegelwel amount-to-take should certainly never be negative - but it's expected for it to sometimes be 0.
Let's say that the initial mem available by role is {"cook" 200 "*" 200"}.
Task 1 needs 300 MB of memory... after it is done the remaining resource pool for memory will be: {"cook" 0  "*" 100 }
So if task 2 only needs 50MB of memory, it will first look at "cook", see that amount-to-take is 0, then move onto  "*" and take 50MB from the remaining 100 there.